content/post/104-dht-deep-dive.md

```diff
@@ -0,0 +1,154 @@
+        *   If a user receives a newer record (i.e., a record that has a higher IPNS sequence number), it updates its existing one and continues until the lookup terminates. This is needed in order to make sure that the user gets the latest record. Recall that IPNS records are mutable and therefore, we need to make sure that we point a request to the latest version of the content.
+            *   The default in go-ipfs will abort early after receiving 16 records, but it can be set to go until the query terminates
+        *   Once the lookup is done, if any of the `K` closest peers to `X` did not have the newest IPNS record, send them the newest record
+*   Peer Records (For a peer where the multihash of the public key is `H`)
+    *   Put: This occurs implicitly - when libp2p peers connect to each other they exchange peer information automatically. Being part of the DHT (as either a client or server) requires being in frequent contact with your `K` closest peers, therefore they will inherently end up with your peer record.
+    *   Get: Do a lookup for the `K` closest peers to `X=SHA256(H)`, but instead of just asking each peer in the lookup “who are the `K` closest peers to `X` you know about?” also ask “please send me the peer record for `H` if you have it”
+        *   We will try to connect to the peer with ID `H` as soon as we learn addresses about it. The lookup can terminate early if we end up connecting to the peer.
+
+## Testing and Results
+
+There have been a lot of changes to the DHT as part of the go-ipfs v0.5.0 release. While many of the changes were intuitively going to be quite useful, we needed harder evidence that the full set of changes would result in a stable and performant network. To that end, we utilized Testground, which is a new distributed testing infrastructure (check out their launch notes at [Testground blog post](https://blog.ipfs.io/2020-05-06-launching-testground/)). 
+
+Throughout the development process we ran many Testground tests to get an understanding of how our changes have improved the network. Below is a comparison of the performance of a 1000 peer network where all peers have around 100-120ms latencies from each other, that is running the DHT from go-ipfs v0.4.23 and the DHT from go-ipfs v0.5.0. A few notes: The v0.4.23 DHT had a few modifications to make testing easier like removing hard coded lookup timeouts, so we see just how long the queries should really be running for. Additionally, while the v0.4.23 test had 60% of the peers be undialable, the v0.5.0 test had 0 undialable nodes. This change actually benefits the v0.4.23 results, since in v0.5.0 if 60% of the nodes are undialable then, because v0.5.0 undialable nodes no longer advertise themselves as DHT servers, the network itself effectively shrinks to 40% of its size, which decreases lookup times.
```

Comment:
> Additionally, while the v0.4.23 test had 60% of the peers be undialable, the v0.5.0 test had 0 undialable nodes. This change actually benefits the v0.4.23 results, since in v0.5.0 if 60% of the nodes are undialable then, because v0.5.0 undialable nodes no longer advertise themselves as DHT servers, the network itself effectively shrinks to 40% of its size, which decreases lookup times.

This is not clear and is a bit hard to follow. It states 0.4.23 had an advantage in the test suite, but the followup statements make it sound like 0.5 had the benefit in testing.

If this isn't easy to clarify succinctly I would just remove it.

Reply:
I'm fine removing it, it was more of a "full disclosure" kind of thing.

The v0.4.23 test had undialable nodes in it and the v0.5.0 did not. That might seem like an advantage to the v0.5.0 test since undialable=bad, but because of our work on v0.5.0 we actually biased the results towards helping v0.4.23 and still trounced it easily.

Reply:
I think removing it is fine since it already mentions needing to tweak 0.4.23 for testing purposes.

Reply:
```suggestion
Throughout the development process we ran many Testground tests to get an understanding of how our changes have improved the network. Below is a comparison of the performance of a 1000 peer network where all peers have around 100-120ms latencies from each other, that is running the DHT from go-ipfs v0.4.23 and the DHT from go-ipfs v0.5.0. A few notes: The v0.4.23 DHT had a few modifications to make testing easier like removing hard coded lookup timeouts, so we see just how long the queries should really be running for.
```